ql/src/java/org/apache/hadoop/hive/ql/metadata/SessionHiveMetaStoreClient.java

```diff
@@ -1201,7 +1201,7 @@ public List<String> listPartitionNames(String catName, String dbName, String tbl
   public List<String> listPartitionNames(String catName, String dbName, String tblName,
       List<String> partVals, int maxParts) throws TException {
     org.apache.hadoop.hive.metastore.api.Table table = getTempTable(dbName, tblName);
-    if (table == null) {
+    if (table == null || partVals == null) {
```

Comment:
why if the table is a temporary table while the `partVals` is null?

Reply:
It would throw MetaException like `HiveMetaStoreClient`.

Reply:
Why not revert back the check on [PartitionTree.java](https://github.com/apache/hive/pull/5206/files#diff-f83827d7c6ec0a73bbc558101e64145afb555daff397c999c7491373253168c7), it's a bit confusing a temporary table just for performing a null check in `HiveMetaStoreClient`

Reply:
Because `SessionHiveMetaStoreClient` extends `HiveMetaStoreClient`, it should also check null for part_vals like `HiveMetaStoreClient`, and `PartitionTree` should support empty or even null part_vals like metastore server in my opinion.

Otherwise it would cause the inconsistent test results in `TestListPartitions` for `HiveMetaStoreClient` and `SessionHiveMetaStoreClient`, for example `testListPartitionsByValuesNoVals()`. The master branch mark this test expect `MetaException`, but actually it should return all partitions for empty part_vals. It could pass the test in master because it throws `MetaException` when making part name for empty partVals in `getNumPartitionsByPs`, which should be a bug in implement this method.

Reply:
> Because SessionHiveMetaStoreClient extends HiveMetaStoreClient, it should also check null for part_vals like HiveMetaStoreClient

I think you cannot just push the null check of a temporary table in `HiveMetaStoreClient`,  and the `PartitionTree` is for the temporary table to process the partition request, it works at the client side, why not put the check here?

Reply:
> the PartitionTree is for the temporary table to process the partition request, it works at the client side, why not put the check here?

Because `get_partitions_ps_with_auth` support null and empty part_vals, the null check is added in client side. So in `PartitionTree` we can also follow the server behavior and push null check in `SessionHiveMetaStoreClient`. This change seems to have no side effects, and in turn supports empty part_vals like `HiveMetaStoreClient`.

Reply:
I don't like this idea: null check for temporary table in HiveMetaStoreClient, let check others' opinion. cc @deniskuzZ @zhangbutao

Reply:
I tried to understand you folk's talk:

@wecharyu thinks that some hms partition api support `null ` & `empty `, such as `get_partitions_ps_with_auth`, but some hms partition api don't support `null ` & `empty`. So, for temp table, check `null `& `empry` and always throw exception in `PartitionTree ` in not inappropriate. 
I think in this case, @wecharyu you need to make sure all the api's behaviors which invoke the `PartitionTree ::getPartitionsByPartitionVals`  are unaffected. For example, `TempTable::getPartitionsByPartitionVals`  is invoked in four api, but i see that two of them are not changed after you change `PartitionTree ::getPartitionsByPartitionVals`, is that safe?


@dengzhhu653 I am not very familair with temp table's ability.  `PartitionTree` is only used for temp table, if temp table has same ability with common table, especially partition ability, and  meanwhile some hms api like `get_partitions_ps_with_auth` can allow `null `& `empty `, i think we can check in `HiveMetaStoreClient` side, and it is good for me to unify the behavior of common table and temp table.

Reply:
> but i see that two of them are not changed

One is that `SessionHiveMetaStoreClient#istPartitionNamesRequest` does not check null for part_vals, I think it's acceptable because `HiveMetaStoreClient#listPartitionNamesRequest` also not check it. 

The other is `SessionHiveMetaStoreClient#exchangePartitions` where part_vals can not be null.

I can not find out the side effect on enable null or empty part_vals in these calls if we do regard null or empty part_vals is used to fetch all parts or part names.